File: extension/res/CanvasAudioClip.java

import java.applet.AudioClip;

public class CanvasAudioClip implements AudioClip {

	private int clipIndex;

	public CanvasAudioClip(URL url){
		clipIndex = constructAudio(url.toString());
	}

	public native void play();
	public native void loop();
	public native void stop();

	private native int constructAudio(String url);

}

File: extension/res/natives/CanvasAudioClip.js

(function(){
	let audioClips = [];

	registerNatives({
	  'CanvasAudioClip': {
	  	'constructAudio(Ljava/lang/String;)I': function(thread, javaThis, arg0){
			let url = arg0.toString();
			let clip = new Audio(url);
			return audioClips.push(clip) - 1;
	  	},
	  	'play()V': function(thread, javaThis){
	  		let index = javaThis['CanvasAudioClip/clipIndex'];
        let clip = audioClips[index];
        clip.play();
	   	},
      'loop()V': function(thread, javaThis){
        let index = javaThis['CanvasAudioClip/clipIndex'];
        let clip = audioClips[index];
        clip.loop = true;
        clip.play();
      },
      'stop()V': function(thread, javaThis){
        let index = javaThis['CanvasAudioClip/clipIndex'];
        let clip = audioClips[index];
      }
	  }
	});
})();
